npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb

```diff
@@ -2,18 +2,213 @@
+      def unsupported?
+        false
+      end
+    end
+
+    class PNPMPackageManager < Ecosystem::VersionManager
+      extend T::Sig
+      NAME = "pnpm"
+      LOCKFILE_NAME = "pnpm-lock.yaml"
+
+      PNPM_V7 = "7"
+      PNPM_V8 = "8"
+      PNPM_V9 = "9"
+
+      SUPPORTED_VERSIONS = T.let([
+        Version.new(PNPM_V7),
+        Version.new(PNPM_V8),
+        Version.new(PNPM_V9)
+      ].freeze, T::Array[Dependabot::Version])
+
+      DEPRECATED_VERSIONS = T.let([].freeze, T::Array[Dependabot::Version])
+
+      sig { params(raw_version: String).void }
+      def initialize(raw_version)
+        super(
+          NAME,
+          Version.new(raw_version),
+          DEPRECATED_VERSIONS,
+          SUPPORTED_VERSIONS
+        )
+      end
+
+      sig { override.returns(T::Boolean) }
+      def deprecated?
+        false
+      end
+
+      sig { override.returns(T::Boolean) }
+      def unsupported?
+        false
+      end
+    end
+
+    PACKAGE_MANAGER_CLASSES = {
+      NpmPackageManager::NAME => NpmPackageManager,
+      YarnPackageManager::NAME => YarnPackageManager,
+      PNPMPackageManager::NAME => PNPMPackageManager
+    }.freeze
+
+    class PackageManagerDetector
+      extend T::Sig
+      extend T::Helpers
+
+      sig do
+        params(
+          lockfiles: T::Hash[Symbol, T.nilable(Dependabot::DependencyFile)],
+          package_json: T::Hash[String, T.untyped]
+        ).void
+      end
+      def initialize(lockfiles, package_json)
+        @lockfiles = lockfiles
+        @package_json = package_json
+        @manifest_package_manager = package_json["packageManager"]
+        @engines = package_json.fetch("engines", nil)
+      end
+
+      sig { returns(T.nilable(String)) }
+      def detect_package_manager
+        name_from_lockfiles || name_from_package_manager_attr || name_from_engines
+      end
+
+      private
+
+      sig { returns(T.nilable(String)) }
+      def name_from_lockfiles
+        # We prioritize the package manager that has a lockfile
+        PACKAGE_MANAGER_CLASSES.each_key do |manager_name| # iterates keys in order as defined in the hash
+          return manager_name.to_s if @lockfiles[manager_name.to_sym]
+        end
+        nil
+      end
+
+      sig { returns(T.nilable(String)) }
+      def name_from_package_manager_attr
+        return unless @manifest_package_manager
+
+        # We prioritize the package manager that has a lockfile
+        PACKAGE_MANAGER_CLASSES.each_key do |manager_name| # iterates keys in order as defined in the hash
+          return manager_name.to_s if @manifest_package_manager.start_with?("#{manager_name}@")
+        end
+      end
+
+      sig { returns(T.nilable(String)) }
+      def name_from_engines
+        return unless @engines.is_a?(Hash)
+
+        PACKAGE_MANAGER_CLASSES.each_key do |manager_name| # iterates keys in order as defined in the hash
+          return manager_name.to_s if @engines[manager_name.to_s]
+        end
+      end
+    end
```

Comment:
This class is created to detect which package manager is going to be used by `dependabot` to run update process.